### PR TITLE
BUG: Fix uint-overflow if padding with linear_ramp and negative gain

### LINKED ISSUE #14191: Unexpected behaviour by numpy.pad() with mode="linear_ramp" and uint dtypes in 1.17
Since 1.17 `numpy.pad()` with `mode="linear_ramp"` seems to not use the array's edge value as the end value, but instead the value range maximum of the array's dtype. This seems to happen only for uint dtypes and only if `end_value > edge_value`.

### Reproducing code example:

```python
import sys, numpy as np; print(np.__version__, sys.version)

for dtype in ["uint8", "uint16", "uint32",
              "int8", "int16", "int32",
              "float16", "float32", "float64"]:
    arr = np.zeros((1,), dtype=dtype)
    arr_pad = np.pad(arr, (3, 0), mode="linear_ramp", end_values=3)
    print("%07s: %s" % (dtype, arr_pad))
```

Output for 1.17:
```bash
1.17.0 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21) 
[GCC 7.3.0]
  uint8: [  3  87 172   0]
 uint16: [    3 21847 43692     0]
 uint32: [         3 1431655767 2863311532          0]
   int8: [3 2 1 0]
  int16: [3 2 1 0]
  int32: [3 2 1 0]
float16: [3. 2. 1. 0.]
float32: [3. 2. 1. 0.]
float64: [3. 2. 1. 0.]
```

Output for 1.16.4 (on a python 2.7 system):
```bash
('1.16.4', '2.7.15 |Anaconda, Inc.| (default, Oct 23 2018, 18:31:10) \n[GCC 7.3.0]')
  uint8: [3 2 1 0]
 uint16: [3 2 1 0]
 uint32: [3 2 1 0]
   int8: [3 2 1 0]
  int16: [3 2 1 0]
  int32: [3 2 1 0]
float16: [3. 2. 1. 0.]
float32: [3. 2. 1. 0.]
float64: [3. 2. 1. 0.]
```

### Further Observations

Using a 2D array does not fix the issue.

Using `end_values=(3, 0)` instead of `end_values=3` does not fix the issue.

However, the issue does not appear in 1.17 so long as the edge value is `>=3`:
```python
import sys, numpy as np; print(np.__version__, sys.version)

for dtype in ["uint8", "uint16", "uint32",
              "int8", "int16", "int32",
              "float16", "float32", "float64"]:
    arr = np.zeros((1,), dtype=dtype)
    arr = arr + 3
    arr_pad = np.pad(arr, (3, 0), mode="linear_ramp", end_values=(3, 0))
    print("%07s: %s" % (dtype, arr_pad))
```
Output:
```bash
1.17.0 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21) 
[GCC 7.3.0]
  uint8: [3 3 3 3]
 uint16: [3 3 3 3]
 uint32: [3 3 3 3]
   int8: [3 3 3 3]
  int16: [3 3 3 3]
  int32: [3 3 3 3]
float16: [3. 3. 3. 3.]
float32: [3. 3. 3. 3.]
float64: [3. 3. 3. 3.]
```

### Related

Maybe this is related to issue #12293?

There is also a paragraph in the 1.17 release notes saying:

> Additionally, there are some corner cases with behavior changes:
> Padding max < min has changed to be more consistent across dtypes, but should not be relied upon.

This is in a section about changes to `clip`, but maybe it is still somehow related?

### PR DESCRIPTION
**Pull request by lagru**
In case end_value was greater than the edge values of the array to pad, the resulting difference was negative and caused overflows in case the input array used unsigned types.  Fixed by using 1.17 version of `linspace` instead of `_linear_ramp`.

Protect against this regression with a new test for all numeric dtypes.

Fixes #14191.

cc @seberg, @aleju

### CODE REVIEW

**Comment by lagru** [comment]
CI failures seem unrelated.

**Comment by lagru** [comment]
I'm not sure how the "09 - Backport-Candidate" label is understood here but simply cherry-picking these changes won't work on the 1.16-release because of the rewrite of pad's internals in the meantime...

**Comment by seberg** [comment]
Oh, sorry. I thought this was a regression. It is a bug that exists seperately both on 1.16 and 1.7 I guess.

File: numpy/lib/arraypad.py
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: You might instead want to use `np.lib.histograms._unsigned_subtract`, possibly after moving it to a more sensible shared location

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Although looking again, that doesn't work too well

File: numpy/lib/arraypad.py
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Perhaps use `ramp.dtype` instead of `np.float16`?

Comment (lagru) [reply numpy/lib/arraypad.py]: That's a good idea. I took it further by finding the most fitting type at the start of the function which is then used everywhere. That might save us a few surprises down the line in case the function will be used elsewhere or "start" and "stop" are swapped.

Off topic: I wish I had known about the existence of np.result_type and similar functions earlier. They seem very useful. :sweat_smile:

**Comment by mattip** [comment]
@seberg ping

**Comment by charris** [comment]
@eric-wieser Ping.

**Comment by eric-wieser** [comment]
Will take a look when I'm back from vacation next week

**Comment by charris** [comment]
Ping all :)

File: numpy/lib/arraypad.py
Comment (seberg) [inline numpy/lib/arraypad.py]: Thanks for the ping chuck. The `np.float16` seems still a bit strange to me. Can we make this `result_type` include final output? That way you get high precision floating point if necessary. I realize that you do not have that dtype available here naturally. An alternative may be to cast the input values early on as well.

Of course that might still fail if everything is unsigned, hmmm.

The current `np.float16` is not trustworthy, a scalar float could be better. This way if start and stop are scalars, the result dtype will actually end up being float16, and we do not want that happening.

Comment (lagru) [reply numpy/lib/arraypad.py]: What do you mean by "final output"? In the current implementation the parameter `stop` is the slice of the edge of the padded array and has the same dtype as the returned array.

What do you mean by "scalar float"? This?
```python
np.result_type(float, start, stop)
```

Comment (seberg) [reply numpy/lib/arraypad.py]: Oh, I mean `np.result_type(start, stop, 0.)`, final output is the input array dtype, yes. I think it should rather be `np.result_type(start, stop, input_dtype)` maybe, although if all are uint, things are different.

**Comment by charris** [comment]
Pushing off to 1.17.2. Will still put in in 1.17.1 if done in time.

File: numpy/lib/arraypad.py
Comment (lagru) [reply numpy/lib/arraypad.py]: Didn't notice the edge cases between `float` and `0.` as parameters. I think `np.result_type(start, stop, 0.)` fulfills all our constraints.

Thought: Maybe we should move the statement `_round_if_needed(right_ramp, padded.dtype)` (called right after this function) into it. That would consolidate the dtype management into one place.

**Comment by charris** [comment]
I note that there is already a backport for 1.16.5. I think that should be closed until this is ready and merged.

File: numpy/lib/arraypad.py
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Perhaps just `Number` instead of naming `float` and `int` - arrays of `Decimal` should work here too, and the numpy complex types I'd also expect to work.

File: numpy/lib/arraypad.py
Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Can we just call `linspace` here, and leave it to handle the corner cases?

Comment (lagru) [reply numpy/lib/arraypad.py]: Huh, linspace supports non-scalars since 1.16. Good to know! Your suggestion doesn't work because the reshaping doesn't translate to the output of linspace. However I'll see how I can use linspace here and how it compares.

Comment (lagru) [reply numpy/lib/arraypad.py]: Found a solution and moved that one into _get_linear_ramps and deleted the obsolete _linear_ramp function. Haven't yet benchmarked it.

File: numpy/lib/arraypad.py
Comment (seberg) [inline numpy/lib/arraypad.py]: Looks good to me, I would not worry about the speed, at least for now. I do not quite understand why the `if` is needed above, as far as I can tell the linspace output must be 2D? (you may also just use `right_ramp.squeeze(axis + 1)`)

Comment (lagru) [reply numpy/lib/arraypad.py]: Good catch about the unnecessary if-statement!

File: numpy/lib/arraypad.py
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Does any of this docstring need transferring to below?

File: numpy/lib/arraypad.py
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: I don't think the `+1` works correctly if `axis` is negative.

If you squeeze before linspace, I think the problem will go away - `start=end_value_pair[1].squeeze(axis)` etc

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Actually, I'm now wondering if we should add a `keepdims=True` to linspace, with semantics "replace the `axis`th axis with the new linspace and squeeze". @mhvk, do you have thoughts on this?

Comment (mhvk) [reply numpy/lib/arraypad.py]: @eric-wieser - you mean something like the opposite of a reduction, where `keepdims=True` tells `linspace` that there is already a dimension of `1` inside `start` and `stop`? On the one hand, it doesn't quite feel as natural, since in that case one has to check that that dimension is indeed `1`. On the other hand, I can see the nice aspect that then `start` and `stop` would broadcast with the result. Overall, the "other hand" wins for me...

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Exactly.

File: numpy/lib/arraypad.py
Comment (lagru) [reply numpy/lib/arraypad.py]: In my opinion, no. All parameters of `_get_linear_ramps` are already documented and `_linear_ramp` is fully replaced by `linspace`. Oh, and `edge_pair` is documented as the return value of `_get_edges`. Should be fine.

File: numpy/lib/arraypad.py
Comment (lagru) [reply numpy/lib/arraypad.py]: > If you squeeze before linspace, I think the problem will go away - start=end_value_pair[1].squeeze(axis) etc

That did the trick. I think broadcasting the end_value_pair is not necessary. linspace does that itself when constructing `delta`. Testing seems to confirm that.

**Comment by charris** [comment]
@eric-wieser Ping.

File: numpy/lib/arraypad.py
Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Opened #14425 from that suggestion

**Review by eric-wieser — APPROVED** [review]
Really pleased that the linspace improvements allowed a bunch of code to be deleted here!

**Comment by charris** [comment]
Thanks @lagru . Time for #14211

**Comment by lagru** [comment]
You're welcome but the important ideas where all yours. :wink:

**Pull request merged**
